### PR TITLE
`Equivalent` trait

### PR DESCRIPTION
**Pull request by timothee-haudebourg**
This PR introduces an `Equivalent` trait, as requested in #345, to customize lookup operations without requiring a `Borrow` implementation. It provides a blanket implementation that covers the `Borrow` case so it should not break anything.

The `hash_map::EntryRef` API, although it uses the `Borrow` trait, remains unchanged as this would introduce some breaking changes.

### CODE REVIEW

**Comment by timothee-haudebourg** [comment]
I said that the `EntryRef` API is unchanged, but the `HashMap::entry_ref` method uses the `Equivalent` trait. Maybe it would be better to continue using the `Borrow` trait for this method, since the resulting `EntryRef` will probably be unusable without a `Borrow` implementation anyway.

**Review by Amanieu — APPROVED** [review]
LGTM! Let me know once you have fixed the typo so I can merge this.

**Comment by Amanieu** [comment]
ping @timothee-haudebourg, is this ready to merge?

**Comment by timothee-haudebourg** [comment]
Sorry I forgot about this, I just pushed some typo fixes, it should be ready to merge!

**Comment by Amanieu** [comment]
Could you add a test for this? Perhaps just copy one from indexmap: https://github.com/bluss/indexmap/blob/master/tests/equivalent_trait.rs

**Comment by Amanieu** [comment]
@bors r+

Thanks!

**Comment by dhedey** [comment]
Just to give a heads up for anyone else hitting this problem.

When I upgraded `HashBrown` past 1.3.1 which released this feature, I got some compile errors on code I'd previously written. The error was:
```
the trait `Borrow<&X>` is not implemented for `X`
= note: required for `&X` to implement `hashbrown::Equivalent<X>`
```

It transpires that we previously had double references: `let x = &abc; hashmap.get(&x)` - and the code previously was happy with them, but the inference with the blanket impl of `Equivalent` isn't as clever as with the direct `Borrow` impl.

I could fix this in my code by removing the extra `&` (or equivalently, adding `*` before the reference).

----

Separately, from a hashbrown compatibility perspective, is this an issue? For us at least it means the build passed fine using `std::HashMap`, but broke using `hashbrown` as a stand in replacement when building for no-std.

**Comment by mrcnski** [comment]
Thanks @dhedey, we hit the same error, and simply removing an extra `&` worked to resolve it!

**Comment by mattklein123** [comment]
@Amanieu sorry for the late drive by on this PR: would you accept a PR which makes the default implementation of Equivalent gated behind a feature which defaults to on? I have a use case where I would like to implement Equivalent on top of a generic structure and I can't figure out a way to prevent this from conflicting with the default implementation. Thank you!

**Comment by Amanieu** [comment]
Note that in the next release we will use the trait from the `equivalent` crate instead (https://github.com/rust-lang/hashbrown/pull/442). Would this help in your case?

**Comment by Amanieu** [comment]
Unfortunately I don't think we can disable the `Equivalent` trait: it's needed by `HashMap` to support `Borrow`.

**Comment by mattklein123** [comment]
> Note that in the next release we will use the trait from the equivalent crate instead (https://github.com/rust-lang/hashbrown/pull/442). Would this help in your case?

I'm not sure. I can take a look once it releases.

> Unfortunately I don't think we can disable the Equivalent trait: it's needed by HashMap to support Borrow.

I wasn't suggesting disabling the equivalent trait, I was thinking more that we could allow the blanket implementation to be disabled and force the user to implement it however they want? It's the blanket implementation that is getting me.

**Comment by Amanieu** [comment]
Can you show the impl you are trying to write? I would think the Rust coherence rules would always prevent you from implementing a conflicting impl.

**Comment by mattklein123** [comment]
> Can you show the impl you are trying to write? I would think the Rust coherence rules would always prevent you from implementing a conflicting impl.

Let me try to come up with a working example when I have some time. Basically I just want this code removed from my build: https://github.com/rust-lang/hashbrown/blob/40ab9e6cb9458f2920ff40530b42245db0c12787/src/lib.rs#L149-L157

The coherence rules are preventing me from implementing a conflicting impl. :)

**Comment by Amanieu** [comment]
I believe the coherence rules would prevent you regardless of whether this impl exists in hashbrown. The problem is that your crate doesn't "own" the trait, so you cannot write blanket impls.

**Pull request merged**
